File: scripts/parse_v2_semantic_judge.py
import json
from collections import defaultdict
from pathlib import Path

path = Path("testdata/evals/rewrite/v2_24_semantic_judge.json")
raw = path.read_bytes()
idx = raw.rfind(b'{"suite"')
if idx < 0:
    raise SystemExit("no suite json in output")
text = raw[idx:].decode("utf-8", errors="replace").lstrip("\ufeff")
data = json.loads(text)
samples = data["samples"]
agg = data["aggregate"]

print("RUN", data["run_metadata"]["run_at"])
passed = sum(1 for s in samples if s["passed"])
print(f"RULE_PASS {passed}/{len(samples)} ({agg['pass_rate']*100:.1f}%)")
metrics = agg.get("metrics") or {}
print(f"MRR uplift {metrics.get('mrr_uplift', 0):+.3f}")
print(f"avg_semantic {metrics.get('avg_semantic_score')}")
print(f"avg_judge {metrics.get('avg_judge_score')}")
print(f"semantic_judge_overrides {metrics.get('semantic_judge_override_count', 0)}")

by_path = defaultdict(int)
rule_only_pass = 0
for s in samples:
    path = (s.get("scores") or {}).get("pass_path")
    if path:
        by_path[path] += 1
    if s["passed"] and (s.get("rule_checks") or {}).get("rule_passed"):
        rule_only_pass += 1
print("\nPASS_PATH", dict(by_path))
print(f"passed_with_rule_green {rule_only_pass}/{passed}")

rule_fail_judge_pass = []
rule_fail_semantic_pass = []
rule_fail_both_pass = []
for s in samples:
    rule_pass = s["passed"]
    scores = s.get("scores") or {}
    judge = scores.get("judge_score")
    semantic = scores.get("semantic_score")
    judge_pass = judge is not None and judge >= 0.65
    semantic_pass = semantic is not None and semantic >= 0.65
    if not rule_pass and judge_pass:
        rule_fail_judge_pass.append((s["name"], judge, semantic))
    if not rule_pass and semantic_pass:
        rule_fail_semantic_pass.append(s["name"])
    if not rule_pass and judge_pass and semantic_pass:
        rule_fail_both_pass.append(s["name"])

print(f"\nRULE_FAIL but JUDGE>=0.65: {len(rule_fail_judge_pass)}")
for name, j, sem in rule_fail_judge_pass:
    print(f"  {name}: judge={j}, semantic={sem}")

print(f"\nRULE_FAIL but SEMANTIC>=0.65: {len(rule_fail_semantic_pass)}")
for name in rule_fail_semantic_pass:
    print(f"  {name}")

print(f"\nRULE_FAIL but BOTH>=0.65: {len(rule_fail_both_pass)}")
for name in rule_fail_both_pass:
    print(f"  {name}")

print("\nRULE_FAILURES")
for s in samples:
    if not s["passed"]:
        print(f"  {s['name']}: {s.get('failure_reasons')}")

File: scripts/parse_v2_semantic_judge.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

func main() {
	raw, err := os.ReadFile("testdata/evals/rewrite/v2_24_soft_gate.json")
	if err != nil {
		panic(err)
	}
	idx := -1
	marker := []byte(`{"suite"`)
	for i := len(raw) - len(marker); i >= 0; i-- {
		if string(raw[i:i+len(marker)]) == string(marker) {
			idx = i
			break
		}
	}
	if idx < 0 {
		panic("suite json not found")
	}
	var data struct {
		RunMetadata struct {
			RunAt string `json:"run_at"`
		} `json:"run_metadata"`
		Samples []struct {
			Name           string         `json:"name"`
			Passed         bool           `json:"passed"`
			FailureReasons []string       `json:"failure_reasons"`
			Scores         map[string]any `json:"scores"`
			RuleChecks     map[string]any `json:"rule_checks"`
		} `json:"samples"`
		Aggregate struct {
			PassRate float64 `json:"pass_rate"`
			Metrics  map[string]any `json:"metrics"`
		} `json:"aggregate"`
	}
	if err := json.Unmarshal(raw[idx:], &data); err != nil {
		panic(err)
	}
	passed := 0
	var ruleFailJudgePass, ruleFailSemanticPass, ruleFailBoth []string
	for _, s := range data.Samples {
		if s.Passed {
			passed++
			continue
		}
		judge, _ := s.Scores["judge_score"].(float64)
		semantic, _ := s.Scores["semantic_score"].(float64)
		if judge >= 0.65 {
			ruleFailJudgePass = append(ruleFailJudgePass, fmt.Sprintf("%s (judge=%.2f semantic=%.2f)", s.Name, judge, semantic))
		}
		if semantic >= 0.65 {
			ruleFailSemanticPass = append(ruleFailSemanticPass, s.Name)
		}
		if judge >= 0.65 && semantic >= 0.65 {
			ruleFailBoth = append(ruleFailBoth, s.Name)
		}
	}
	fmt.Println("RUN", data.RunMetadata.RunAt)
	fmt.Printf("RULE_PASS %d/%d (%.1f%%)\n", passed, len(data.Samples), data.Aggregate.PassRate*100)
	fmt.Printf("MRR uplift %v\n", data.Aggregate.Metrics["mrr_uplift"])
	fmt.Printf("avg_semantic %v\n", data.Aggregate.Metrics["avg_semantic_score"])
	fmt.Printf("semantic_judge_overrides %v\n", data.Aggregate.Metrics["semantic_judge_override_count"])

	byPath := map[string]int{}
	ruleOnlyPass := 0
	for _, s := range data.Samples {
		if path, ok := s.Scores["pass_path"].(string); ok && path != "" {
			byPath[path]++
		}
		if s.Passed {
			if rulePassed, ok := s.RuleChecks["rule_passed"].(bool); ok && rulePassed {
				ruleOnlyPass++
			}
		}
	}
	fmt.Printf("\nPASS_PATH %v\n", byPath)
	fmt.Printf("passed_with_rule_green %d/%d\n", ruleOnlyPass, passed)
	fmt.Printf("\nRULE_FAIL but JUDGE>=0.65: %d\n", len(ruleFailJudgePass))
	for _, line := range ruleFailJudgePass {
		fmt.Println(" ", line)
	}
	fmt.Printf("\nRULE_FAIL but SEMANTIC>=0.65: %d\n", len(ruleFailSemanticPass))
	for _, name := range ruleFailSemanticPass {
		fmt.Println(" ", name)
	}
	fmt.Printf("\nRULE_FAIL but BOTH>=0.65: %d\n", len(ruleFailBoth))
	for _, name := range ruleFailBoth {
		fmt.Println(" ", name)
	}
	fmt.Println("\nRULE_FAILURES")
	for _, s := range data.Samples {
		if !s.Passed {
			fmt.Printf("  %s: %v\n", s.Name, s.FailureReasons)
		}
	}
}
